// react-antd-business-components/lib/chart/PieChart.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const recharts = require("recharts");
const RechartsTooltip = require("./Tooltip");
const VarietyChart = require("./VarietyChart");
const jsxRuntime = require("react/jsx-runtime");
const _interopDefaultLegacy = (e) => e && typeof e === "object" && "default" in e ? e : { default: e };
const RechartsTooltip__default = /* @__PURE__ */ _interopDefaultLegacy(RechartsTooltip);
const PieChart = (props) => {
  var _a;
  const {
    width = 250,
    height = 250,
    data: outerData = [],
    formatter = (v) => v,
    valuePrecision = 0,
    percentPrecision = 0,
    useResponsiveContainer,
    showLabel,
    outerRadius,
    innerRadius,
    ...restProps
  } = props;
  const lastFormatter = (value) => formatter(Number(value.toFixed(valuePrecision))).toLocaleString();
  let data = [].concat(outerData).filter(Boolean);
  if (!((_a = data[0]) == null ? void 0 : _a.chartData)) {
    data = [{
      chartData: data
    }];
  }
  const chartContent = /* @__PURE__ */ jsxRuntime.jsxs(recharts.PieChart, {
    width,
    height,
    ...restProps,
    children: [/* @__PURE__ */ jsxRuntime.jsx(recharts.Tooltip, {
      content: /* @__PURE__ */ jsxRuntime.jsx(RechartsTooltip__default.default, {}),
      formatter: lastFormatter
    }), data.map((d, index) => {
      const {
        chartData = [],
        ...rest
      } = d;
      const dataKey = rest.dataKey || "value";
      const allValueCounts = chartData.reduce((counts, currentData) => counts + currentData[dataKey], 0);
      const lastColors = VarietyChart.colors.slice();
      chartData.forEach((c) => {
        if (c.fill) {
          lastColors.unshift(c.fill);
        }
        c.percent = c[dataKey] / allValueCounts;
      });
      let label = ({
        name,
        value,
        percent
      }) => {
        const lastValue = lastFormatter(value);
        if (rest.label === "percent") {
          return `${(percent * 100).toFixed(percentPrecision)}%`;
        }
        if (rest.label === "name") {
          return name;
        }
        if (rest.label === "value") {
          return lastValue;
        }
        return `${name} (${(percent * 100).toFixed(percentPrecision)}%)`;
      };
      if (rest.label === false || !showLabel) {
        label = false;
      }
      return /* @__PURE__ */ jsxRuntime.jsx(recharts.Pie, {
        outerRadius,
        innerRadius,
        fill: lastColors[index],
        ...rest,
        dataKey,
        data: chartData,
        label,
        animationDuration: 500,
        isAnimationActive: false,
        children: chartData.map((_, index2) => /* @__PURE__ */ jsxRuntime.jsx(recharts.Cell, {
          fill: lastColors[index2]
        }, `cell-${index2}`))
      }, index.toString());
    })]
  });
  if (useResponsiveContainer) {
    return /* @__PURE__ */ jsxRuntime.jsx("div", {
      style: {
        height
      },
      children: /* @__PURE__ */ jsxRuntime.jsx(recharts.ResponsiveContainer, {
        width: "100%",
        height,
        children: chartContent
      })
    });
  }
  return /* @__PURE__ */ jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: chartContent
  });
};
PieChart.displayName = "PieChart";
PieChart.defaultProps = {
  useResponsiveContainer: true,
  height: 300,
  showLabel: true,
  valuePrecision: 2,
  percentPrecision: 2
};
exports.default = PieChart;
