// react-antd-business-components/dist/rbac-components.es.js
import { Space as Or, Table as Ar, Alert as Re, Empty as Ge } from "antd";
import Ne, { useMemo as ge, useState as Sr, useEffect as jr } from "react";
import ye from "classnames";
import { isNumber as kr, getFormatValue as Fe, getStringLength as Je, getStringDOMWidth as Se } from "business-utils";
import { Symbols as Dr, Surface as $r, AreaChart as Nr, Area as Xe, BarChart as Fr, Bar as He, ComposedChart as Ir, Line as je, LineChart as Lr, CartesianGrid as Vr, XAxis as Wr, YAxis as Br, Tooltip as ar, Brush as Yr, Legend as Kr, ResponsiveContainer as nr, PieChart as Mr, Pie as Ur, Cell as zr } from "recharts";
import { getNiceTickValues as qr } from "recharts-scale";
var xe = { exports: {} }, oe = {};
/*
object-assign
(c) Sindre Sorhus
@license MIT
*/
var ke, Ze;
function ir() {
  if (Ze)
    return ke;
  Ze = 1;
  var i = Object.getOwnPropertySymbols, p = Object.prototype.hasOwnProperty, u = Object.prototype.propertyIsEnumerable;
  function o(c) {
    if (c == null)
      throw new TypeError("Object.assign cannot be called with null or undefined");
    return Object(c);
  }
  function _() {
    try {
      if (!Object.assign)
        return !1;
      var c = new String("abc");
      if (c[5] = "de", Object.getOwnPropertyNames(c)[0] === "5")
        return !1;
      for (var d = {}, a = 0; a < 10; a++)
        d["_" + String.fromCharCode(a)] = a;
      var h = Object.getOwnPropertyNames(d).map(function(n) {
        return d[n];
      });
      if (h.join("") !== "0123456789")
        return !1;
      var s = {};
      return "abcdefghijklmnopqrst".split("").forEach(function(n) {
        s[n] = n;
      }), Object.keys(Object.assign({}, s)).join("") === "abcdefghijklmnopqrst";
    } catch {
      return !1;
    }
  }
  return ke = _() ? Object.assign : function(c, d) {
    for (var a, h = o(c), s, n = 1; n < arguments.length; n++) {
      a = Object(arguments[n]);
      for (var v in a)
        p.call(a, v) && (h[v] = a[v]);
      if (i) {
        s = i(a);
        for (var g = 0; g < s.length; g++)
          u.call(a, s[g]) && (h[s[g]] = a[s[g]]);
      }
    }
    return h;
  }, ke;
}
/** @license React v17.0.2
 * react-jsx-runtime.production.min.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var Qe;
function Gr() {
  if (Qe)
    return oe;
  Qe = 1, ir();
  var i = Ne, p = 60103;
  if (oe.Fragment = 60107, typeof Symbol == "function" && Symbol.for) {
    var u = Symbol.for;
    p = u("react.element"), oe.Fragment = u("react.fragment");
  }
  var o = i.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED.ReactCurrentOwner, _ = Object.prototype.hasOwnProperty, c = { key: !0, ref: !0, __self: !0, __source: !0 };
  function d(a, h, s) {
    var n, v = {}, g = null, m = null;
    s !== void 0 && (g = "" + s), h.key !== void 0 && (g = "" + h.key), h.ref !== void 0 && (m = h.ref);
    for (n in h)
      _.call(h, n) && !c.hasOwnProperty(n) && (v[n] = h[n]);
    if (a && a.defaultProps)
      for (n in h = a.defaultProps, h)
        v[n] === void 0 && (v[n] = h[n]);
    return { $$typeof: p, type: a, key: g, ref: m, props: v, _owner: o.current };
  }
  return oe.jsx = d, oe.jsxs = d, oe;
}
var De = {};
/** @license React v17.0.2
 * react-jsx-runtime.development.js
 *
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
var er;
function Jr() {
  return er || (er = 1, function(i) {
    process.env.NODE_ENV !== "production" && function() {
      var p = Ne, u = ir(), o = 60103, _ = 60106;
      i.Fragment = 60107;
      var c = 60108, d = 60114, a = 60109, h = 60110, s = 60112, n = 60113, v = 60120, g = 60115, m = 60116, w = 60121, L = 60122, N = 60117, Y = 60129, U = 60131;
      if (typeof Symbol == "function" && Symbol.for) {
        var C = Symbol.for;
        o = C("react.element"), _ = C("react.portal"), i.Fragment = C("react.fragment"), c = C("react.strict_mode"), d = C("react.profiler"), a = C("react.provider"), h = C("react.context"), s = C("react.forward_ref"), n = C("react.suspense"), v = C("react.suspense_list"), g = C("react.memo"), m = C("react.lazy"), w = C("react.block"), L = C("react.server.block"), N = C("react.fundamental"), C("react.scope"), C("react.opaque.id"), Y = C("react.debug_trace_mode"), C("react.offscreen"), U = C("react.legacy_hidden");
      }
      var V = typeof Symbol == "function" && Symbol.iterator, ce = "@@iterator";
      function z(e) {
        if (e === null || typeof e != "object")
          return null;
        var r = V && e[V] || e[ce];
        return typeof r == "function" ? r : null;
      }
      var K = p.__SECRET_INTERNALS_DO_NOT_USE_OR_YOU_WILL_BE_FIRED;
      function R(e) {
        {
          for (var r = arguments.length, t = new Array(r > 1 ? r - 1 : 0), l = 1; l < r; l++)
            t[l - 1] = arguments[l];
          q("error", e, t);
        }
      }
      function q(e, r, t) {
        {
          var l = K.ReactDebugCurrentFrame, x = l.getStackAddendum();
          x !== "" && (r += "%s", t = t.concat([x]));
          var T = t.map(function(b) {
            return "" + b;
          });
          T.unshift("Warning: " + r), Function.prototype.apply.call(console[e], console, T);
        }
      }
      var Q = !1;
      function ue(e) {
        return !!(typeof e == "string" || typeof e == "function" || e === i.Fragment || e === d || e === Y || e === c || e === n || e === v || e === U || Q || typeof e == "object" && e !== null && (e.$$typeof === m || e.$$typeof === g || e.$$typeof === a || e.$$typeof === h || e.$$typeof === s || e.$$typeof === N || e.$$typeof === w || e[0] === L));
      }
      function fe(e, r, t) {
        var l = r.displayName || r.name || "";
        return e.displayName || (l !== "" ? t + "(" + l + ")" : t);
      }
      function be(e) {
        return e.displayName || "Context";
      }
      function F(e) {
        if (e == null)
          return null;
        if (typeof e.tag == "number" && R("Received an unexpected object in getComponentName(). This is likely a bug in React. Please file an issue."), typeof e == "function")
          return e.displayName || e.name || null;
        if (typeof e == "string")
          return e;
        switch (e) {
          case i.Fragment:
            return "Fragment";
          case _:
            return "Portal";
          case d:
            return "Profiler";
          case c:
            return "StrictMode";
          case n:
            return "Suspense";
          case v:
            return "SuspenseList";
        }
        if (typeof e == "object")
          switch (e.$$typeof) {
            case h:
              var r = e;
              return be(r) + ".Consumer";
            case a:
              var t = e;
              return be(t._context) + ".Provider";
            case s:
              return fe(e, e.render, "ForwardRef");
            case g:
              return F(e.type);
            case w:
              return F(e._render);
            case m: {
              var l = e, x = l._payload, T = l._init;
              try {
                return F(T(x));
              } catch {
                return null;
              }
            }
          }
        return null;
      }
      var H = 0, de, E, j, k, S, W, ee;
      function re() {
      }
      re.__reactDisabledLog = !0;
      function te() {
        {
          if (H === 0) {
            de = console.log, E = console.info, j = console.warn, k = console.error, S = console.group, W = console.groupCollapsed, ee = console.groupEnd;
            var e = {
              configurable: !0,
              enumerable: !0,
              value: re,
              writable: !0
            };
            Object.defineProperties(console, {
              info: e,
              log: e,
              warn: e,
              error: e,
              group: e,
              groupCollapsed: e,
              groupEnd: e
            });
          }
          H++;
        }
      }
      function P() {
        {
          if (H--, H === 0) {
            var e = {
              configurable: !0,
              enumerable: !0,
              writable: !0
            };
            Object.defineProperties(console, {
              log: u({}, e, {
                value: de
              }),
              info: u({}, e, {
                value: E
              }),
              warn: u({}, e, {
                value: j
              }),
              error: u({}, e, {
                value: k
              }),
              group: u({}, e, {
                value: S
              }),
              groupCollapsed: u({}, e, {
                value: W
              }),
              groupEnd: u({}, e, {
                value: ee
              })
            });
          }
          H < 0 && R("disabledDepth fell below zero. This is a bug in React. Please file an issue.");
        }
      }
      var I = K.ReactCurrentDispatcher, G;
      function D(e, r, t) {
        {
          if (G === void 0)
            try {
              throw Error();
            } catch (x) {
              var l = x.stack.trim().match(/\n( *(at )?)/);
              G = l && l[1] || "";
            }
          return `
` + G + e;
        }
      }
      var pe = !1, ae;
      {
        var Z = typeof WeakMap == "function" ? WeakMap : Map;
        ae = new Z();
      }
      function _e(e, r) {
        if (!e || pe)
          return "";
        {
          var t = ae.get(e);
          if (t !== void 0)
            return t;
        }
        var l;
        pe = !0;
        var x = Error.prepareStackTrace;
        Error.prepareStackTrace = void 0;
        var T;
        T = I.current, I.current = null, te();
        try {
          if (r) {
            var b = function() {
              throw Error();
            };
            if (Object.defineProperty(b.prototype, "props", {
              set: function() {
                throw Error();
              }
            }), typeof Reflect == "object" && Reflect.construct) {
              try {
                Reflect.construct(b, []);
              } catch (X) {
                l = X;
              }
              Reflect.construct(e, [], b);
            } else {
              try {
                b.call();
              } catch (X) {
                l = X;
              }
              e.call(b.prototype);
            }
          } else {
            try {
              throw Error();
            } catch (X) {
              l = X;
            }
            e();
          }
        } catch (X) {
          if (X && l && typeof X.stack == "string") {
            for (var y = X.stack.split(`
`), $ = l.stack.split(`
`), O = y.length - 1, A = $.length - 1; O >= 1 && A >= 0 && y[O] !== $[A]; )
              A--;
            for (; O >= 1 && A >= 0; O--, A--)
              if (y[O] !== $[A]) {
                if (O !== 1 || A !== 1)
                  do
                    if (O--, A--, A < 0 || y[O] !== $[A]) {
                      var J = `
` + y[O].replace(" at new ", " at ");
                      return typeof e == "function" && ae.set(e, J), J;
                    }
                  while (O >= 1 && A >= 0);
                break;
              }
          }
        } finally {
          pe = !1, I.current = T, P(), Error.prepareStackTrace = x;
        }
        var ie = e ? e.displayName || e.name : "", qe = ie ? D(ie) : "";
        return typeof e == "function" && ae.set(e, qe), qe;
      }
      function he(e, r, t) {
        return _e(e, !1);
      }
      function ur(e) {
        var r = e.prototype;
        return !!(r && r.isReactComponent);
      }
      function Ce(e, r, t) {
        if (e == null)
          return "";
        if (typeof e == "function")
          return _e(e, ur(e));
        if (typeof e == "string")
          return D(e);
        switch (e) {
          case n:
            return D("Suspense");
          case v:
            return D("SuspenseList");
        }
        if (typeof e == "object")
          switch (e.$$typeof) {
            case s:
              return he(e.render);
            case g:
              return Ce(e.type, r, t);
            case w:
              return he(e._render);
            case m: {
              var l = e, x = l._payload, T = l._init;
              try {
                return Ce(T(x), r, t);
              } catch {
              }
            }
          }
        return "";
      }
      var Ie = {}, Le = K.ReactDebugCurrentFrame;
      function Ee(e) {
        if (e) {
          var r = e._owner, t = Ce(e.type, e._source, r ? r.type : null);
          Le.setExtraStackFrame(t);
        } else
          Le.setExtraStackFrame(null);
      }
      function fr(e, r, t, l, x) {
        {
          var T = Function.call.bind(Object.prototype.hasOwnProperty);
          for (var b in e)
            if (T(e, b)) {
              var y = void 0;
              try {
                if (typeof e[b] != "function") {
                  var $ = Error((l || "React class") + ": " + t + " type `" + b + "` is invalid; it must be a function, usually from the `prop-types` package, but received `" + typeof e[b] + "`.This often happens because of typos such as `PropTypes.function` instead of `PropTypes.func`.");
                  throw $.name = "Invariant Violation", $;
                }
                y = e[b](r, b, l, t, null, "SECRET_DO_NOT_PASS_THIS_OR_YOU_WILL_BE_FIRED");
              } catch (O) {
                y = O;
              }
              y && !(y instanceof Error) && (Ee(x), R("%s: type specification of %s `%s` is invalid; the type checker function must return `null` or an `Error` but returned a %s. You may have forgotten to pass an argument to the type checker creator (arrayOf, instanceOf, objectOf, oneOf, oneOfType, and shape all require an argument).", l || "React class", t, b, typeof y), Ee(null)), y instanceof Error && !(y.message in Ie) && (Ie[y.message] = !0, Ee(x), R("Failed %s type: %s", t, y.message), Ee(null));
            }
        }
      }
      var me = K.ReactCurrentOwner, Te = Object.prototype.hasOwnProperty, dr = {
        key: !0,
        ref: !0,
        __self: !0,
        __source: !0
      }, Ve, We, Pe;
      Pe = {};
      function pr(e) {
        if (Te.call(e, "ref")) {
          var r = Object.getOwnPropertyDescriptor(e, "ref").get;
          if (r && r.isReactWarning)
            return !1;
        }
        return e.ref !== void 0;
      }
      function hr(e) {
        if (Te.call(e, "key")) {
          var r = Object.getOwnPropertyDescriptor(e, "key").get;
          if (r && r.isReactWarning)
            return !1;
        }
        return e.key !== void 0;
      }
      function mr(e, r) {
        if (typeof e.ref == "string" && me.current && r && me.current.stateNode !== r) {
          var t = F(me.current.type);
          Pe[t] || (R('Component "%s" contains the string ref "%s". Support for string refs will be removed in a future major release. This case cannot be automatically converted to an arrow function. We ask you to manually fix this case by using useRef() or createRef() instead. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-string-ref', F(me.current.type), e.ref), Pe[t] = !0);
        }
      }
      function vr(e, r) {
        {
          var t = function() {
            Ve || (Ve = !0, R("%s: `key` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
          };
          t.isReactWarning = !0, Object.defineProperty(e, "key", {
            get: t,
            configurable: !0
          });
        }
      }
      function gr(e, r) {
        {
          var t = function() {
            We || (We = !0, R("%s: `ref` is not a prop. Trying to access it will result in `undefined` being returned. If you need to access the same value within the child component, you should pass it as a different prop. (https://reactjs.org/link/special-props)", r));
          };
          t.isReactWarning = !0, Object.defineProperty(e, "ref", {
            get: t,
            configurable: !0
          });
        }
      }
      var yr = function(e, r, t, l, x, T, b) {
        var y = {
          $$typeof: o,
          type: e,
          key: r,
          ref: t,
          props: b,
          _owner: T
        };
        return y._store = {}, Object.defineProperty(y._store, "validated", {
          configurable: !1,
          enumerable: !1,
          writable: !0,
          value: !1
        }), Object.defineProperty(y, "_self", {
          configurable: !1,
          enumerable: !1,
          writable: !1,
          value: l
        }), Object.defineProperty(y, "_source", {
          configurable: !1,
          enumerable: !1,
          writable: !1,
          value: x
        }), Object.freeze && (Object.freeze(y.props), Object.freeze(y)), y;
      };
      function br(e, r, t, l, x) {
        {
          var T, b = {}, y = null, $ = null;
          t !== void 0 && (y = "" + t), hr(r) && (y = "" + r.key), pr(r) && ($ = r.ref, mr(r, x));
          for (T in r)
            Te.call(r, T) && !dr.hasOwnProperty(T) && (b[T] = r[T]);
          if (e && e.defaultProps) {
            var O = e.defaultProps;
            for (T in O)
              b[T] === void 0 && (b[T] = O[T]);
          }
          if (y || $) {
            var A = typeof e == "function" ? e.displayName || e.name || "Unknown" : e;
            y && vr(b, A), $ && gr(b, A);
          }
          return yr(e, y, $, x, l, me.current, b);
        }
      }
      var we = K.ReactCurrentOwner, Be = K.ReactDebugCurrentFrame;
      function ne(e) {
        if (e) {
          var r = e._owner, t = Ce(e.type, e._source, r ? r.type : null);
          Be.setExtraStackFrame(t);
        } else
          Be.setExtraStackFrame(null);
      }
      var Oe;
      Oe = !1;
      function Ae(e) {
        return typeof e == "object" && e !== null && e.$$typeof === o;
      }
      function Ye() {
        {
          if (we.current) {
            var e = F(we.current.type);
            if (e)
              return `

Check the render method of \`` + e + "`.";
          }
          return "";
        }
      }
      function _r(e) {
        {
          if (e !== void 0) {
            var r = e.fileName.replace(/^.*[\\\/]/, ""), t = e.lineNumber;
            return `

Check your code at ` + r + ":" + t + ".";
          }
          return "";
        }
      }
      var Ke = {};
      function Cr(e) {
        {
          var r = Ye();
          if (!r) {
            var t = typeof e == "string" ? e : e.displayName || e.name;
            t && (r = `

Check the top-level render call using <` + t + ">.");
          }
          return r;
        }
      }
      function Me(e, r) {
        {
          if (!e._store || e._store.validated || e.key != null)
            return;
          e._store.validated = !0;
          var t = Cr(r);
          if (Ke[t])
            return;
          Ke[t] = !0;
          var l = "";
          e && e._owner && e._owner !== we.current && (l = " It was passed a child from " + F(e._owner.type) + "."), ne(e), R('Each child in a list should have a unique "key" prop.%s%s See https://reactjs.org/link/warning-keys for more information.', t, l), ne(null);
        }
      }
      function Ue(e, r) {
        {
          if (typeof e != "object")
            return;
          if (Array.isArray(e))
            for (var t = 0; t < e.length; t++) {
              var l = e[t];
              Ae(l) && Me(l, r);
            }
          else if (Ae(e))
            e._store && (e._store.validated = !0);
          else if (e) {
            var x = z(e);
            if (typeof x == "function" && x !== e.entries)
              for (var T = x.call(e), b; !(b = T.next()).done; )
                Ae(b.value) && Me(b.value, r);
          }
        }
      }
      function Er(e) {
        {
          var r = e.type;
          if (r == null || typeof r == "string")
            return;
          var t;
          if (typeof r == "function")
            t = r.propTypes;
          else if (typeof r == "object" && (r.$$typeof === s || r.$$typeof === g))
            t = r.propTypes;
          else
            return;
          if (t) {
            var l = F(r);
            fr(t, e.props, "prop", l, e);
          } else if (r.PropTypes !== void 0 && !Oe) {
            Oe = !0;
            var x = F(r);
            R("Component %s declared `PropTypes` instead of `propTypes`. Did you misspell the property assignment?", x || "Unknown");
          }
          typeof r.getDefaultProps == "function" && !r.getDefaultProps.isReactClassApproved && R("getDefaultProps is only used on classic React.createClass definitions. Use a static property named `defaultProps` instead.");
        }
      }
      function Rr(e) {
        {
          for (var r = Object.keys(e.props), t = 0; t < r.length; t++) {
            var l = r[t];
            if (l !== "children" && l !== "key") {
              ne(e), R("Invalid prop `%s` supplied to `React.Fragment`. React.Fragment can only have `key` and `children` props.", l), ne(null);
              break;
            }
          }
          e.ref !== null && (ne(e), R("Invalid attribute `ref` supplied to `React.Fragment`."), ne(null));
        }
      }
      function ze(e, r, t, l, x, T) {
        {
          var b = ue(e);
          if (!b) {
            var y = "";
            (e === void 0 || typeof e == "object" && e !== null && Object.keys(e).length === 0) && (y += " You likely forgot to export your component from the file it's defined in, or you might have mixed up default and named imports.");
            var $ = _r(x);
            $ ? y += $ : y += Ye();
            var O;
            e === null ? O = "null" : Array.isArray(e) ? O = "array" : e !== void 0 && e.$$typeof === o ? (O = "<" + (F(e.type) || "Unknown") + " />", y = " Did you accidentally export a JSX literal instead of a component?") : O = typeof e, R("React.jsx: type is invalid -- expected a string (for built-in components) or a class/function (for composite components) but got: %s.%s", O, y);
          }
          var A = br(e, r, t, x, T);
          if (A == null)
            return A;
          if (b) {
            var J = r.children;
            if (J !== void 0)
              if (l)
                if (Array.isArray(J)) {
                  for (var ie = 0; ie < J.length; ie++)
                    Ue(J[ie], e);
                  Object.freeze && Object.freeze(J);
                } else
                  R("React.jsx: Static children should always be an array. You are likely explicitly calling React.jsxs or React.jsxDEV. Use the Babel transform instead.");
              else
                Ue(J, e);
          }
          return e === i.Fragment ? Rr(A) : Er(A), A;
        }
      }
      function xr(e, r, t) {
        return ze(e, r, t, !0);
      }
      function Tr(e, r, t) {
        return ze(e, r, t, !1);
      }
      var Pr = Tr, wr = xr;
      i.jsx = Pr, i.jsxs = wr;
    }();
  }(De)), De;
}
(function(i) {
  process.env.NODE_ENV === "production" ? i.exports = Gr() : i.exports = Jr();
})(xe);
const f = xe.exports.jsx, M = xe.exports.jsxs, se = xe.exports.Fragment, Xr = (i) => {
  const {
    children: p,
    direction: u
  } = i;
  let {
    style: o = {}
  } = i;
  return u === "vertical" && (o = {
    ...o,
    width: "100%",
    display: "flex"
  }), /* @__PURE__ */ f(Or, {
    ...i,
    direction: u,
    style: o,
    children: p
  });
};
Xr.displayName = "RSpace";
const Hr = "rabc-table__container", Zr = "rabc-table__td-no-wrap", Qr = "rabc-table__td-wrap", rr = {
  container: Hr,
  tdNoWrap: Zr,
  tdWrap: Qr
}, or = (i) => {
  var m;
  const {
    className: p,
    scrollY: u,
    showPagination: o,
    rowKey: _,
    emptyFieldValue: c,
    pagination: d,
    columns: a = [],
    scroll: h,
    dataSource: s
  } = i, n = ge(() => et({
    columns: a,
    emptyFieldValue: c,
    styles: rr
  }), [a, c]);
  let v = _;
  !_ && ((m = s == null ? void 0 : s[0]) == null ? void 0 : m.id) !== void 0 && (v = "id");
  let g = d;
  return o === !1 && (g = !1), /* @__PURE__ */ f(Ar, {
    ...i,
    scroll: {
      x: !0,
      y: u,
      ...h
    },
    className: ye(p, rr.container),
    dataSource: s,
    rowKey: v,
    pagination: g,
    columns: n
  });
};
or.displayName = "RTable";
or.defaultProps = {
  showPagination: !1,
  emptyFieldValue: "-"
};
function et(i) {
  const {
    columns: p,
    emptyFieldValue: u,
    styles: o = {}
  } = i, _ = [], c = [], d = [];
  return p.filter((a) => !a.hide).forEach((a) => {
    const {
      render: h
    } = a, s = {
      ...a,
      wrap: !!a.width,
      title: /* @__PURE__ */ f("div", {
        className: o.th,
        style: {
          width: a.width
        },
        children: a.title
      }),
      render(n, v, g) {
        var w;
        if (n == null && !a.notDealEmptyValue)
          return u;
        let m = ((w = a.formatter) == null ? void 0 : w.call(a, n)) || n;
        return (typeof m == "string" || kr(m)) && typeof a.format == "string" && (m = Fe(m, a.format)), h && (m = h(m, v, g)), /* @__PURE__ */ f("div", {
          style: {
            width: a.width
          },
          children: m
        });
      }
    };
    s.wrap ? (s.className = ye(a.className, o.tdWrap), delete s.wrap) : s.className = ye(a.className, o.tdNoWrap), s.fixed === !0 || s.fixed === "left" ? _.push(s) : s.fixed === "right" ? c.push(s) : d.push(s);
  }), _.concat(d).concat(c);
}
const sr = (i) => {
  const {
    showSearchTip: p,
    searchTip: u,
    showNoDataTip: o,
    noDataTip: _,
    showWarnTip: c,
    warnTip: d,
    showErrorTip: a,
    errorTip: h,
    showSuccessTip: s,
    successTip: n
  } = i;
  return a ? /* @__PURE__ */ f(Re, {
    type: "error",
    message: h,
    showIcon: !0
  }) : c ? /* @__PURE__ */ f(Re, {
    type: "warning",
    message: d,
    showIcon: !0
  }) : s ? /* @__PURE__ */ f(Re, {
    type: "success",
    message: n,
    showIcon: !0
  }) : o ? /* @__PURE__ */ f(Ge, {
    image: Ge.PRESENTED_IMAGE_SIMPLE,
    description: _
  }) : p ? /* @__PURE__ */ f(Re, {
    type: "info",
    message: u,
    showIcon: !0
  }) : /* @__PURE__ */ f(se, {
    children: i.children
  });
};
sr.displayName = "ProContent";
sr.defaultProps = {
  searchTip: "\u8BF7\u5728\u4E0A\u65B9\u8F93\u5165\u7136\u540E\u70B9\u51FB\u67E5\u8BE2",
  noDataTip: "\u6682\u65E0\u6570\u636E",
  successTip: "\u6210\u529F\u63D0\u793A",
  warnTip: "\u8B66\u544A\u63D0\u793A",
  errorTip: "\u9519\u8BEF\u63D0\u793A"
};
const rt = "rabc-chart-tooltip__recharts-tooltip", tt = "rabc-chart-tooltip__rechart-tooltip-label", at = "rabc-chart-tooltip__recharts-tooltip-content", nt = "rabc-chart-tooltip__rechart-tooltip-item-value", ve = {
  rechartsTooltip: rt,
  rechartTooltipLabel: tt,
  rechartsTooltipContent: at,
  rechartTooltipItemValue: nt
};
function lr(i) {
  const {
    active: p
  } = i;
  if (p) {
    const {
      isPercent: u,
      payload: o = [],
      label: _,
      formatter: c,
      unit: d,
      titleColor: a,
      numberColor: h
    } = i;
    let s = 0;
    return u && (s = o.reduce((n, v) => n + Number(v.value), 0)), /* @__PURE__ */ M("div", {
      className: ve.rechartsTooltip,
      children: [_ && /* @__PURE__ */ M("div", {
        className: ve.rechartTooltipLabel,
        style: {
          color: a
        },
        children: [`${_}`, u && /* @__PURE__ */ M(se, {
          children: ["\xA0(\u603B\u6570\uFF1A", s.toLocaleString(), ")"]
        })]
      }), /* @__PURE__ */ f("ul", {
        className: ve.rechartsTooltipContent,
        children: o == null ? void 0 : o.map((n, v) => {
          let g = d;
          const {
            percent: m
          } = n.payload, w = c ? c(n.value) : n.value;
          return (n.value === 0 || n.value === "N/A") && (g = null), /* @__PURE__ */ M("li", {
            children: [/* @__PURE__ */ M("span", {
              className: ve.rechartTooltipItemTitle,
              style: {
                color: h || n.stroke
              },
              children: [n.name, "\uFF1A"]
            }), /* @__PURE__ */ M("span", {
              className: ve.rechartTooltipItemValue,
              children: [w.toLocaleString(), g, u && /* @__PURE__ */ M(se, {
                children: ["\xA0(", `${(n.value / (s || 1) * 100).toFixed(2)}%`, ")"]
              }), m && /* @__PURE__ */ M(se, {
                children: ["\xA0(", `${(m * 100).toFixed(2)}%`, ")"]
              })]
            })]
          }, v);
        })
      })]
    });
  }
  return null;
}
const B = 32, tr = "rgba(0,0,0,.25)";
class it extends Ne.Component {
  renderIcon(p) {
    const {
      inactiveColor: u
    } = this.props, o = B / 2, _ = B / 6, c = B / 3, d = (p.inactive ? u : p.color) || tr;
    return p.type === "plainline" ? /* @__PURE__ */ f("line", {
      strokeWidth: 4,
      fill: "none",
      stroke: d,
      strokeDasharray: p.payload.strokeDasharray,
      x1: 0,
      y1: o,
      x2: B,
      y2: o,
      className: "recharts-legend-icon"
    }) : p.type === "line" ? /* @__PURE__ */ f("path", {
      strokeWidth: 4,
      fill: "none",
      stroke: d,
      d: `M0,${o}h${c}
            A${_},${_},0,1,1,${2 * c},${o}
            H${B}M${2 * c},${o}
            A${_},${_},0,1,1,${c},${o}`,
      className: "recharts-legend-icon"
    }) : p.type === "rect" ? /* @__PURE__ */ f("path", {
      stroke: "none",
      fill: d,
      d: `M0,${B / 8}h${B}v${B * 3 / 4}h${-B}z`,
      className: "recharts-legend-icon"
    }) : /* @__PURE__ */ f(Dr, {
      fill: d,
      cx: o,
      cy: o,
      size: B,
      sizeType: "diameter",
      type: p.type
    });
  }
  renderItems() {
    const {
      payload: p = [],
      customPayload: u = p,
      iconSize: o,
      layout: _,
      onClick: c,
      formatter: d
    } = this.props, a = {
      x: 0,
      y: 0,
      width: B,
      height: B
    }, h = {
      display: _ === "horizontal" ? "inline-block" : "block",
      marginRight: 10
    }, s = {
      display: "inline-block",
      verticalAlign: "middle",
      marginRight: 4
    }, n = p.reduce((v, g) => ({
      ...v,
      [g.dataKey]: g
    }), {});
    return u.map((v, g) => {
      const m = n[v.dataKey] || v, w = m.formatter || d, L = ye({
        "recharts-legend-item": !0,
        [`legend-item-${g}`]: !0,
        inactive: m.inactive
      });
      return m.type === "none" ? null : /* @__PURE__ */ M("li", {
        className: L,
        style: {
          ...h,
          cursor: "pointer"
        },
        onClick: () => {
          c == null || c(m);
        },
        children: [/* @__PURE__ */ f($r, {
          width: o,
          height: o,
          viewBox: a,
          style: s,
          children: this.renderIcon(m)
        }), /* @__PURE__ */ f("span", {
          className: "recharts-legend-item-text",
          style: {
            color: m.color ? "" : tr
          },
          children: w ? w(m.value, m, g) : m.value
        })]
      }, `legend-item-${g}`);
    });
  }
  render() {
    const {
      payload: p,
      customPayload: u = p,
      layout: o,
      align: _
    } = this.props;
    return !u || !u.length ? null : /* @__PURE__ */ f("ul", {
      className: "recharts-default-legend",
      style: {
        padding: 0,
        margin: 0,
        textAlign: o === "horizontal" ? _ : "left"
      },
      children: this.renderItems()
    });
  }
}
function ot(i) {
  const { stackOffset: p, layout: u, xAxisFormatter: o = (n) => n, yAxisFormatter: _ = (n) => n } = i;
  let { xAxisFormat: c = "", yAxisFormat: d = "" } = i;
  const a = (n, v) => Fe(n, v), h = u === "vertical";
  return p === "expand" && (h ? c = "percent(0)" : d = "percent(0)"), (n) => h ? a(o(n), c || "thousandSemicolon") : a(_(n), d || "thousandSemicolon");
}
const st = "rabc-chart-variety-chart__container", lt = {
  container: st
}, $e = ["#1890ff", "#13c2c2", "#52c41a", "#a0d911", "#fadb14", "#faad14", "#f5222d", "#fa541c", "#fa8c16", "#ffbf00", "#f56a00", "#f7a20f", "#108ee9", "#00a2ae", "#eb2f96", "#722ed1", "#2f54eb"];
function le(i) {
  const {
    type: p,
    xAxisFormatter: u = (E) => E,
    yAxisFormatter: o = (E) => E,
    width: _,
    height: c,
    layout: d,
    stackOffset: a,
    brush: h,
    useResponsiveContainer: s,
    data: {
      chartData: n = [],
      dataKeys: v = [],
      dataKeys: g
    } = {},
    ...m
  } = i;
  let {
    xAxisFormat: w = "",
    yAxisFormat: L = ""
  } = i;
  const N = d === "vertical", [Y, U] = Sr(v);
  jr(() => {
    U(g || []);
  }, [g]), a === "expand" && (N ? w = "percent(0)" : L = "percent(0)");
  const V = ot({
    layout: d,
    stackOffset: a,
    xAxisFormat: w,
    yAxisFormat: L,
    xAxisFormatter: u,
    yAxisFormatter: o
  });
  let {
    ChartContainer: ce,
    ChartItem: z
  } = ge(() => p === "area" ? {
    ChartContainer: Nr,
    ChartItem: Xe
  } : p === "bar" ? {
    ChartContainer: Fr,
    ChartItem: He
  } : p === "composed" ? {
    ChartContainer: Ir,
    ChartItem: je
  } : {
    ChartContainer: Lr,
    ChartItem: je
  }, [p]);
  const {
    lastChartData: K,
    margin: R,
    yTicks: q
  } = ge(() => {
    const E = {
      top: 0,
      left: -52,
      right: 0,
      bottom: 6
    }, j = (P) => Fe(P, N ? L : w), k = (P) => N ? o(j(P)) : u(j(P)), S = [], W = [], ee = Y.reduce((P, I) => (I.hide || (P[I.dataKey] = !0), P), {}), re = n.map((P) => {
      const I = k(P.name), G = {
        ...P
      };
      if (N)
        W.push(I);
      else
        for (const D in P)
          ee[D] && (S.push(Number(P[D])), P[D] !== null && (G[D] = Number(P[D])));
      return {
        ...G,
        name: I
      };
    });
    let te;
    if (N) {
      const P = W.reduce((G, D) => Je(String(D)) > Je(String(G)) ? D : G, ""), I = Se(P);
      E.left = E.left + I;
    } else if (a === "expand")
      E.left = E.left + Se("100%");
    else {
      const P = Math.max.apply(null, S), I = (() => {
        const Z = Math.min.apply(null, S);
        return Z > 0 ? 0 : Z;
      })(), D = qr([I, P], 5).reduce((Z, _e) => {
        const he = String(V(_e));
        return he.length > Z.length ? he : Z;
      }, "");
      I === 1 && P === 1 && (te = [0, 1, 2, 3, 4]);
      const pe = String(V(D)), ae = Se(pe);
      E.left = E.left + ae;
    }
    return {
      yTicks: te,
      lastChartData: re,
      margin: E
    };
  }, [n, d, a, Y]), {
    xAxisProps: Q,
    yAxisProps: ue,
    tooltipProps: fe
  } = ge(() => {
    let E = {}, j = {}, k = {};
    if (d === "vertical") {
      j = {
        dataKey: "name",
        type: "category",
        padding: {
          top: 16,
          bottom: 16
        }
      }, E = {
        type: "number",
        tickFormatter: V,
        padding: {
          left: 16,
          right: 16
        }
      };
      const S = w.includes("percent") && a === "expand";
      k = {
        formatter: S ? null : V,
        isPercent: S
      };
    } else {
      E = {
        dataKey: "name",
        padding: {
          left: 16,
          right: 16
        }
      }, j = {
        tickFormatter: V,
        padding: {
          top: 16,
          bottom: 16
        },
        ticks: q
      };
      const S = L.includes("percent") && a === "expand";
      k = {
        formatter: S ? null : V,
        isPercent: S
      };
    }
    return {
      xAxisProps: E,
      yAxisProps: j,
      tooltipProps: k
    };
  }, [d, w, L]), be = ge(() => v.map((E) => ({
    value: E.name || E.dataKey,
    type: "line",
    ...E
  })), [v]), F = Y.reduce((E, j) => {
    let k = E;
    return j.hide || (k += 1), k;
  }, 0), H = (E) => {
    const {
      dataKey: j
    } = E, k = Y.map((S) => {
      const W = {
        ...S
      };
      return S.dataKey === j && (F !== 1 ? W.hide = !W.hide : W.hide = !1), W;
    });
    U(k);
  }, de = /* @__PURE__ */ M(ce, {
    width: _ || 500,
    height: c || 300,
    data: K,
    layout: d,
    margin: R,
    stackOffset: a,
    ...m,
    className: ye(lt.container, m.className),
    children: [/* @__PURE__ */ f(Vr, {
      strokeDasharray: "3 3"
    }), F > 0 && /* @__PURE__ */ f(Wr, {
      ...Q
    }), F > 0 && /* @__PURE__ */ f(Br, {
      ...ue
    }), /* @__PURE__ */ f(ar, {
      content: /* @__PURE__ */ f(lr, {
        isPercent: fe.isPercent
      }),
      ...fe,
      formatter: fe.formatter
    }), h && /* @__PURE__ */ f(Yr, {}), v.length > 1 && /* @__PURE__ */ f(Kr, {
      iconType: "line",
      content: /* @__PURE__ */ f(it, {
        customPayload: be,
        onClick: H
      }),
      wrapperStyle: h ? {
        bottom: -2
      } : {
        bottom: 0
      }
    }), Y.map((E, j) => {
      const {
        dataKey: k,
        chartType: S,
        stroke: W = $e[j],
        fill: ee = $e[j],
        hide: re,
        ...te
      } = E;
      return re ? !1 : (p === "composed" && (S === "line" ? z = je : S === "area" ? z = Xe : S === "bar" && (z = He)), /* @__PURE__ */ f(z, {
        type: "monotone",
        dataKey: k,
        stroke: W,
        fill: ee,
        stackId: a === "expand" ? "1" : void 0,
        animationDuration: 500,
        ...te
      }, k));
    })]
  });
  return s ? /* @__PURE__ */ f("div", {
    style: {
      height: c
    },
    children: /* @__PURE__ */ f(nr, {
      width: "100%",
      height: c,
      children: de
    })
  }) : /* @__PURE__ */ f(se, {
    children: de
  });
}
le.displayName = "VarietyChart";
le.defaultProps = {
  layout: "horizontal",
  useResponsiveContainer: !0,
  height: 300,
  brush: !1
};
const ct = (i) => /* @__PURE__ */ f(le, {
  ...i,
  type: "line"
});
ct.displayName = "LineChart";
const ut = (i) => /* @__PURE__ */ f(le, {
  ...i,
  type: "area"
});
ut.displayName = "Area";
const ft = (i) => /* @__PURE__ */ f(le, {
  ...i,
  type: "bar"
});
ft.displayName = "Bar";
const dt = (i) => /* @__PURE__ */ f(le, {
  ...i,
  type: "composed"
});
dt.displayName = "ComposedChart";
const cr = (i) => {
  var L;
  const {
    width: p = 250,
    height: u = 250,
    data: o = [],
    formatter: _ = (N) => N,
    valuePrecision: c = 0,
    percentPrecision: d = 0,
    useResponsiveContainer: a,
    showLabel: h,
    outerRadius: s,
    innerRadius: n,
    ...v
  } = i, g = (N) => _(Number(N.toFixed(c))).toLocaleString();
  let m = [].concat(o).filter(Boolean);
  (L = m[0]) != null && L.chartData || (m = [{
    chartData: m
  }]);
  const w = /* @__PURE__ */ M(Mr, {
    width: p,
    height: u,
    ...v,
    children: [/* @__PURE__ */ f(ar, {
      content: /* @__PURE__ */ f(lr, {}),
      formatter: g
    }), m.map((N, Y) => {
      const {
        chartData: U = [],
        ...C
      } = N, V = C.dataKey || "value", ce = U.reduce((R, q) => R + q[V], 0), z = $e.slice();
      U.forEach((R) => {
        R.fill && z.unshift(R.fill), R.percent = R[V] / ce;
      });
      let K = ({
        name: R,
        value: q,
        percent: Q
      }) => {
        const ue = g(q);
        return C.label === "percent" ? `${(Q * 100).toFixed(d)}%` : C.label === "name" ? R : C.label === "value" ? ue : `${R} (${(Q * 100).toFixed(d)}%)`;
      };
      return (C.label === !1 || !h) && (K = !1), /* @__PURE__ */ f(Ur, {
        outerRadius: s,
        innerRadius: n,
        fill: z[Y],
        ...C,
        dataKey: V,
        data: U,
        label: K,
        animationDuration: 500,
        isAnimationActive: !1,
        children: U.map((R, q) => /* @__PURE__ */ f(zr, {
          fill: z[q]
        }, `cell-${q}`))
      }, Y.toString());
    })]
  });
  return a ? /* @__PURE__ */ f("div", {
    style: {
      height: u
    },
    children: /* @__PURE__ */ f(nr, {
      width: "100%",
      height: u,
      children: w
    })
  }) : /* @__PURE__ */ f(se, {
    children: w
  });
};
cr.displayName = "PieChart";
cr.defaultProps = {
  useResponsiveContainer: !0,
  height: 300,
  showLabel: !0,
  valuePrecision: 2,
  percentPrecision: 2
};
export {
  ut as AreaChart,
  ft as BarChart,
  dt as ComposedChart,
  ct as LineChart,
  cr as PieChart,
  sr as ProContent,
  Xr as Space,
  or as Table,
  le as VarietyChart
};
