// business-utils/lib/index.js
"use strict";
Object.defineProperties(exports, { __esModule: { value: true }, [Symbol.toStringTag]: { value: "Module" } });
const createFullscreen = require("./createFullscreen");
const toCamelCase = require("./toCamelCase");
const isPlainObject = require("./isPlainObject");
const getSearchParams = require("./getSearchParams");
const isNumber = require("./isNumber");
const getStringDOMWidth = require("./getStringDOMWidth");
const getStringLength = require("./getStringLength");
const getFormatConfig = require("./getFormatConfig");
const getFormatValue = require("./getFormatValue");
const _interopDefaultLegacy = (e) => e && typeof e === "object" && "default" in e ? e : { default: e };
const createFullscreen__default = /* @__PURE__ */ _interopDefaultLegacy(createFullscreen);
const toCamelCase__default = /* @__PURE__ */ _interopDefaultLegacy(toCamelCase);
const isPlainObject__default = /* @__PURE__ */ _interopDefaultLegacy(isPlainObject);
const getSearchParams__default = /* @__PURE__ */ _interopDefaultLegacy(getSearchParams);
const isNumber__default = /* @__PURE__ */ _interopDefaultLegacy(isNumber);
const getStringDOMWidth__default = /* @__PURE__ */ _interopDefaultLegacy(getStringDOMWidth);
const getStringLength__default = /* @__PURE__ */ _interopDefaultLegacy(getStringLength);
const getFormatConfig__default = /* @__PURE__ */ _interopDefaultLegacy(getFormatConfig);
const getFormatValue__default = /* @__PURE__ */ _interopDefaultLegacy(getFormatValue);
Object.defineProperty(exports, "createFullscreen", {
  enumerable: true,
  get: () => createFullscreen__default.default
});
Object.defineProperty(exports, "toCamelCase", {
  enumerable: true,
  get: () => toCamelCase__default.default
});
Object.defineProperty(exports, "isPlainObject", {
  enumerable: true,
  get: () => isPlainObject__default.default
});
Object.defineProperty(exports, "getSearchParams", {
  enumerable: true,
  get: () => getSearchParams__default.default
});
Object.defineProperty(exports, "isNumber", {
  enumerable: true,
  get: () => isNumber__default.default
});
Object.defineProperty(exports, "getStringDOMWidth", {
  enumerable: true,
  get: () => getStringDOMWidth__default.default
});
Object.defineProperty(exports, "getStringLength", {
  enumerable: true,
  get: () => getStringLength__default.default
});
Object.defineProperty(exports, "getFomatConfig", {
  enumerable: true,
  get: () => getFormatConfig__default.default
});
Object.defineProperty(exports, "getFormatValue", {
  enumerable: true,
  get: () => getFormatValue__default.default
});
