// react-antd-business-components/es/ProContent/index.js
import { Alert, Empty } from "antd";
import { jsx, Fragment } from "react/jsx-runtime";
const ProContent = (props) => {
  const {
    showSearchTip,
    searchTip,
    showNoDataTip,
    noDataTip,
    showWarnTip,
    warnTip,
    showErrorTip,
    errorTip,
    showSuccessTip,
    successTip
  } = props;
  if (showErrorTip) {
    return /* @__PURE__ */ jsx(Alert, {
      type: "error",
      message: errorTip,
      showIcon: true
    });
  }
  if (showWarnTip) {
    return /* @__PURE__ */ jsx(Alert, {
      type: "warning",
      message: warnTip,
      showIcon: true
    });
  }
  if (showSuccessTip) {
    return /* @__PURE__ */ jsx(Alert, {
      type: "success",
      message: successTip,
      showIcon: true
    });
  }
  if (showNoDataTip) {
    return /* @__PURE__ */ jsx(Empty, {
      image: Empty.PRESENTED_IMAGE_SIMPLE,
      description: noDataTip
    });
  }
  if (showSearchTip) {
    return /* @__PURE__ */ jsx(Alert, {
      type: "info",
      message: searchTip,
      showIcon: true
    });
  }
  return /* @__PURE__ */ jsx(Fragment, {
    children: props.children
  });
};
ProContent.displayName = "ProContent";
ProContent.defaultProps = {
  searchTip: "\u8BF7\u5728\u4E0A\u65B9\u8F93\u5165\u7136\u540E\u70B9\u51FB\u67E5\u8BE2",
  noDataTip: "\u6682\u65E0\u6570\u636E",
  successTip: "\u6210\u529F\u63D0\u793A",
  warnTip: "\u8B66\u544A\u63D0\u793A",
  errorTip: "\u9519\u8BEF\u63D0\u793A"
};
export {
  ProContent as default
};
